contracts/core/modules/IssuanceOrderModule.sol

```diff
@@ -405,48 +327,194 @@ contract IssuanceOrderModule is
+            scannedBytes = scannedBytes.add(exchangeDataLength);
+            makerTokenUsed = makerTokenUsed.add(neededMakerTokenAmount);
+        }
+
+        return makerTokenUsed;
+    }
+
+    /**
+     * Calls exchange to execute trades and deposits fills into Vault for issuanceOrder maker.
+     *
+     * maker                            Issuance order maker
+     * taker                            Issuance order taker
+     * makerToken                       Address of maker token used in exchange orders
+     * makerAssetAmount                 Amount of issuance order maker token to use on this exchange
+     * orderCount                       Expected number of orders to execute
+     * fillQuantity                     Quantity of Set to be filled
+     * attemptedfillQuantity            Quantity of Set taker attempted to fill
+     *
+     * @param  _addresses               [maker, taker, makerToken]
+     * @param  _values                  [makerAssetAmount, orderCount, fillQuantity, attemptedFillQuantity]
+     * @param  _exchange                Address of exchange wrapper being called
+     * @param  _bodyData                Arbitrary bytes data for orders to be executed on exchange
+     */
+    function callExchange(
+        address[3] _addresses,
+        uint256[4] _values,
+        address _exchange,
+        bytes _bodyData
+    )
+        private
+    {
+        // Call Exchange
+        address[] memory componentFillTokens = new address[](_values[1]);
+        uint256[] memory componentFillAmounts = new uint256[](_values[1]);
+        (componentFillTokens, componentFillAmounts) = IExchangeWrapper(_exchange).exchange(
+            _addresses,
+            _values,
+            _bodyData
         );
 
-        uint256[] memory requiredBalances = new uint256[](_order.requiredComponents.length);
+        // Transfer component tokens from wrapper to vault
+        ICore(core).batchDepositModule(
+            _exchange,
+            _addresses[0],
+            componentFillTokens,
+            componentFillAmounts
+        );        
+    }
 
-        // Calculate amount of maker token required
-        uint256 requiredMakerTokenAmount = OrderLibrary.getPartialAmount(
-            _order.makerTokenAmount,
-            _fillQuantity,
-            _order.quantity
+    /**
+     * Check exchange orders acquire correct amount of tokens and taker doesn't over use
+     * the issuance order maker's tokens
+     *
+     * @param  _order                       IssuanceOrder object containing order params
+     * @param  _requiredBalances            Array of required balances for each component
+                                            after exchange orders are executed
+     * @param  _requiredMakerTokenAmount    Max amount of maker token used to source tokens
+     * @param  _makerTokenAmountUsed        Amount of maker token used to source tokens
+     */
+    function checkRequiredTokenBalances(
```

Comment:
https://softwareengineering.stackexchange.com/questions/166884/should-i-extract-specific-functionality-into-a-function-and-why

The thing stand out to me in the above stackoverflow article is "choose good and precise names for your functions". It's generally a really good practice to take an extra second to think about how you will name your function and what the components of the function will be.

Also - We generally use the terms `assert` or `validate`. Perhaps a better name that reflects what this function does is `assertTokenAcquisitionQuantities`.

Reply:
I really do think there has to be a difference in the way we approach solidity development vs traditional. I get the argument that clarity is very important for security reasons. At the same time, there's a balance between that and writing efficient code in this instance. We have to be ok with the idea that people are going to need to grasp compound concepts. I don't think it's practical at this point to separate out all logic into it's own function, there's some middle ground here.

Reply:
When you are talking about efficient code, I'm assuming you mean gas-efficient code. In that case, I believe that writing well-abstracted and understandable code allows us to achieve that goal. When writing efficient code or optimizing, the first place to start is removing unnecessary or redundant SSTORE opcodes. It becomes extremely easier to identify where we could make these fixes when the code is easier to reason about. 

https://github.com/ethereum/pyethereum/blob/develop/ethereum/opcodes.py
^ Opcodes and cost. If you look at the costs of operations, `SSTORE`  (5,000 - 20k gas) is by far the most expensive. `EXTCODECOPY` (now 700 for external contract loading) and `JUMP` (8 gas) are extremely low cost. When the costs are so low, I prefer clarity over gas costs.

Obviously, we don't want to over-abstract - which is when the code itself is clearer than the function. As someone looking at our existing code, it is not easy to quickly understand what is happening. The side effects of large functions it is that we hit stack too deep errors (as we're cramming too much into a function).

Re: `assert` / `validate`, it seems we prefer validate (e.g. `validateSignature``validateOrder`). The only deviance is `checkSuccess` in ERC20Wrapper. Let's go with `validate` going forward.

P.S. I believe there are numerous places in `IssuanceOrderModule` and the `ExchangeWrappers` where we can save gas through duplicate operations.